sql: implemented pg_shdepend

mnovelodou opened the pull request:
Previously, pg_shdepend was unimplemented,
This was inadequate because this table have been reported queried by
telemetry
To address this, this patch implements shared dependencies for tables,
databases, and pinned user/roles

Release note (sql change): implemented pg_shdepend with shared
dependencies with tables, databases and pinned user/roles

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/68018)
<!-- Reviewable:end -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68018)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mnovelodou and @RichardJCai)

___
*[pkg/sql/pg_catalog.go, line 82 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68018#-Mfx79ds22jsMh2_qFCj:-Mfx79ds22jsMh2_qFCk:b44iy9w) ([raw file](https://github.com/cockroachdb/cockroach/blob/0304252ff4b810bd3575e013a48c2274a31d5ace/pkg/sql/pg_catalog.go#L82)):*
> ```Go
> 
> const (
> 	sharedDependencyOwner PGShDependType = "o"
> ```

it would be useful to comment each of these constants (can quote and link to the postgres docs)
___
*[pkg/sql/pg_catalog.go, line 2346 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68018#-Mfx6GXy4hZMgsMW1vxc:-Mfx6GXy4hZMgsMW1vxd:bckj1ie) ([raw file](https://github.com/cockroachdb/cockroach/blob/0304252ff4b810bd3575e013a48c2274a31d5ace/pkg/sql/pg_catalog.go#L2346)):*
> ```Go
> 
> var pgCatalogShdependTable = virtualSchemaTable{
> 	comment: `shared dependencies (empty - not implemented)
> ```

this comment needs to be updated

can you also make sure that the comments for other tables that you have implemented are updated?

if there's a comment in pkg/sql/vtable/pg_catalog.go, make sure that's updated too
___
*[pkg/sql/pg_catalog.go, line 2398 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68018#-Mfx5xUmDt9BC4YpsnvD:-Mfx5xUmDt9BC4YpsnvE:bl14228) ([raw file](https://github.com/cockroachdb/cockroach/blob/0304252ff4b810bd3575e013a48c2274a31d5ace/pkg/sql/pg_catalog.go#L2398)):*
what does "pinnedRoles" mean? why are they special?

i think you can add a comment that quotes the Postgres docs:

> SHARED_DEPENDENCY_PIN (p)
>
>    There is no dependent object; this type of entry is a signal that the system itself depends on the referenced object, and so that object must never be deleted. Entries of this type are created only by initdb. The columns for the dependent object contain zeroes.
___
*[pkg/sql/pg_catalog.go, line 2408 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68018#-Mfx7Odt9LGqIc0qTSKR:-Mfx7Odt9LGqIc0qTSKS:bsnoxfp) ([raw file](https://github.com/cockroachdb/cockroach/blob/0304252ff4b810bd3575e013a48c2274a31d5ace/pkg/sql/pg_catalog.go#L2408)):*
> ```Go
> 				for _, u := range db.GetPrivileges().Show(privilege.Database) {
> 					if _, ok := pinnedRoles[u.User.Normalized()]; ok {
> 						// Pinned roles are not displayed in this table
> ```

why are these roles being excluded?

you should mention that the pinned roles are handled below since they need the `sharedDependencyPin` deptyoe
___
*[pkg/sql/logictest/testdata/logic_test/pg_catalog, line 4796 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68018#-Mfx7-Ty5d168xNGjrhm:-Mfx7-Ty5d168xNGjrhn:by108wj) ([raw file](https://github.com/cockroachdb/cockroach/blob/0304252ff4b810bd3575e013a48c2274a31d5ace/pkg/sql/logictest/testdata/logic_test/pg_catalog#L4796)):*
> ```
> ) ORDER BY rolname
> ----
> rolname
> ```

why is `rolname` a pinned role?
___
*[pkg/sql/logictest/testdata/logic_test/pg_catalog, line 4824 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68018#-Mfx5i2U9YhQDZ21AQoH:-Mfx5i2U9YhQDZ21AQoI:blcg99n) ([raw file](https://github.com/cockroachdb/cockroach/blob/0304252ff4b810bd3575e013a48c2274a31d5ace/pkg/sql/logictest/testdata/logic_test/pg_catalog#L4824)):*
> ```
> rolname  deptype  datname
> admin    a        sh_db
> root     o        sh_db
> ```

in postgres, there is no row with the `o` deptype. do you know why there's a difference?


<!-- Sent from Reviewable.io -->

mnovelodou reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68018)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @RichardJCai)

___
*[pkg/sql/pg_catalog.go, line 82 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68018#-Mfx79ds22jsMh2_qFCj:-Mg7OsCSCet5Q3dZv99l:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0304252ff4b810bd3575e013a48c2274a31d5ace/pkg/sql/pg_catalog.go#L82)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

it would be useful to comment each of these constants (can quote and link to the postgres docs)
</blockquote></details>

Done.
___
*[pkg/sql/pg_catalog.go, line 2346 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68018#-Mfx6GXy4hZMgsMW1vxc:-Mg7Oxcy4KnCPVPRXLIt:bpettxb) ([raw file](https://github.com/cockroachdb/cockroach/blob/0304252ff4b810bd3575e013a48c2274a31d5ace/pkg/sql/pg_catalog.go#L2346)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this comment needs to be updated

can you also make sure that the comments for other tables that you have implemented are updated?

if there's a comment in pkg/sql/vtable/pg_catalog.go, make sure that's updated too
</blockquote></details>

I will fix it now for pg_shdepend and I will create another PR if needed for other tables.
___
*[pkg/sql/pg_catalog.go, line 2398 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68018#-Mfx5xUmDt9BC4YpsnvD:-Mg7OcWvEt4PHbRaopsi:bfzxn2t) ([raw file](https://github.com/cockroachdb/cockroach/blob/0304252ff4b810bd3575e013a48c2274a31d5ace/pkg/sql/pg_catalog.go#L2398)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

what does "pinnedRoles" mean? why are they special?

i think you can add a comment that quotes the Postgres docs:

> SHARED_DEPENDENCY_PIN (p)
>
>    There is no dependent object; this type of entry is a signal that the system itself depends on the referenced object, and so that object must never be deleted. Entries of this type are created only by initdb. The columns for the dependent object contain zeroes.

</blockquote></details>

I have commented this on the constants for deptype
___
*[pkg/sql/pg_catalog.go, line 2408 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68018#-Mfx7Odt9LGqIc0qTSKR:-Mg7DB8z-18rFrFjbOVr:b-y4ukp3) ([raw file](https://github.com/cockroachdb/cockroach/blob/0304252ff4b810bd3575e013a48c2274a31d5ace/pkg/sql/pg_catalog.go#L2408)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

why are these roles being excluded?

you should mention that the pinned roles are handled below since they need the `sharedDependencyPin` deptyoe
</blockquote></details>

I couldn't find objects depending on pinned roles in this table in postgres, that is basically the reason I am excluding them, documentation doesn't help much on this.
___
*[pkg/sql/logictest/testdata/logic_test/pg_catalog, line 4796 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68018#-Mfx7-Ty5d168xNGjrhm:-Mg7BRK9123Bf7l1izw-:b-ilzb5k) ([raw file](https://github.com/cockroachdb/cockroach/blob/0304252ff4b810bd3575e013a48c2274a31d5ace/pkg/sql/logictest/testdata/logic_test/pg_catalog#L4796)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

why is `rolname` a pinned role?
</blockquote></details>

I used `query T colnames` and I selected `rolname` which is the name of the column, is not actually a pinned role
___
*[pkg/sql/logictest/testdata/logic_test/pg_catalog, line 4824 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68018#-Mfx5i2U9YhQDZ21AQoH:-Mg7BneiEJn-CK-MeZNs:bm08eoz) ([raw file](https://github.com/cockroachdb/cockroach/blob/0304252ff4b810bd3575e013a48c2274a31d5ace/pkg/sql/logictest/testdata/logic_test/pg_catalog#L4824)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

in postgres, there is no row with the `o` deptype. do you know why there's a difference?
</blockquote></details>

If the role is the owner of the object you will see `o` deptype, I have done it for a couple of objects in postgres


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
it looks like one of the tests needs to be updated

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68018#-:-MgHYaeo9OF5HITQRgEk:badgh0w)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mnovelodou, @rafiss, and @RichardJCai)

___
*[pkg/sql/logictest/testdata/logic_test/pg_catalog, line 4824 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68018#-Mfx5i2U9YhQDZ21AQoH:-MgHY5Q95qvaSlb-R7Bh:b-r5y0nw) ([raw file](https://github.com/cockroachdb/cockroach/blob/0304252ff4b810bd3575e013a48c2274a31d5ace/pkg/sql/logictest/testdata/logic_test/pg_catalog#L4824)):*
<details><summary><i>Previously, mnovelodou (Miguel Novelo) wrote…</i></summary><blockquote>

If the role is the owner of the object you will see `o` deptype, I have done it for a couple of objects in postgres
</blockquote></details>

can you explain more? i ran this in postgres and there is no `o` deptype:

```
rafiss@127:postgres> create database sh_db with owner = 'rafiss';
rafiss@127:postgres> use sh_db;
rafiss@127:sh_db> CREATE TABLE sh_table_a(id INT);
 CREATE USER sh_user;
 GRANT SELECT ON sh_table_a TO sh_user;
 CREATE ROLE sh_role;
 GRANT SELECT ON sh_table_a TO sh_role;
 GRANT CONNECT ON DATABASE sh_db TO sh_role;

rafiss@127:sh_db> SELECT
   pg_authid.rolname,
   pg_shdepend.deptype,
   pg_database.datname
 FROM pg_authid
 JOIN pg_shdepend ON pg_authid.oid = pg_shdepend.refobjid
 JOIN pg_class ON pg_shdepend.objid = pg_class.oid
 JOIN pg_database ON pg_database.oid = pg_shdepend.dbid
 WHERE pg_class.relname = 'sh_table_a'
 ;
+-----------+-----------+-----------+
| rolname   | deptype   | datname   |
|-----------+-----------+-----------|
| sh_user   | a         | sh_db     |
| sh_role   | a         | sh_db     |
+-----------+-----------+-----------+
```


<!-- Sent from Reviewable.io -->

rafiss commented on pkg/sql/logictest/testdata/logic_test/pg_catalog:
```diff
@@ -4785,3 +4785,58 @@ SELECT sequenceowner, cache_size FROM pg_sequences WHERE sequencename = 'serial'
+----
+rolname  deptype  datname
+admin    a        sh_db
+root     o        sh_db
```
as we discussed in slack, `root` should not appear here because it is a pinned role

we should keep this test, and also add one more test where a non-pinned user is the owner of the database.

mnovelodou replied:
Done.

Commit:
sql: implemented pg_shdepend

Previously, pg_shdepend was unimplemented,
This was inadequate because this table have been reported queried by
telemetry
To address this, this patch implements shared dependencies for tables,
databases, and pinned user/roles

Release note (sql change): implemented pg_shdepend with shared
dependencies with tables, databases and pinned user/roles

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68018)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mnovelodou, @rafiss, and @RichardJCai)

___
*[pkg/sql/logictest/testdata/logic_test/pg_catalog, line 5141 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68018#-MhVIfZTAeKvm9JrSK_v:-MhVIfZTAeKvm9JrSK_w:bm1ass7) ([raw file](https://github.com/cockroachdb/cockroach/blob/04d39d3b04b9742654845b7dbd0199d90408108e/pkg/sql/logictest/testdata/logic_test/pg_catalog#L5141)):*
> ```
> JOIN pg_class ON pg_shdepend.objid = pg_class.oid
> JOIN pg_database ON pg_database.oid = pg_shdepend.dbid
> WHERE pg_class.relname = 'sh_table_a'
> ```

i think it's also useful to add a test where `root` is the owner, but to make sure it doesn't appear in the results because it is a pinned role


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
this looks good! thanks for taking care of all the details

bors r+

> Reviewed 1 of 2 files at r3, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68018#-:-Mh_x8g67MtW2obq_S3O:b8c2vs4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mnovelodou, @rafiss, and @RichardJCai)



<!-- Sent from Reviewable.io -->
